registration/include/pcl/registration/gicp.h

```diff
@@ -375,15 +375,15 @@ class GeneralizedIterativeClosestPoint
 
   /** \return trace of mat1^t . mat2
    * \param mat1 matrix of dimension nxm
-   * \param mat2 matrix of dimension nxp
+   * \param mat2 matrix of dimension nxm
    */
   inline double
   matricesInnerProd(const Eigen::MatrixXd& mat1, const Eigen::MatrixXd& mat2) const
   {
     double r = 0.;
-    std::size_t n = mat1.rows();
+    std::size_t n = mat1.rows(), m = mat1.cols();
     // tr(mat1^t.mat2)
-    for (std::size_t i = 0; i < n; i++)
+    for (std::size_t i = 0; i < m; i++)
       for (std::size_t j = 0; j < n; j++)
         r += mat1(j, i) * mat2(i, j);
```

Comment:
Thoughts on using Eigen?

`(mat1.transpose() * mat2).trace()`

(I think this is the same, and the assembly is also quite similar: https://godbolt.org/z/zYocT7PvM, though the eigen assembly seems to be operating on packed struct, while the loop works on single doubles)

Reply:
It would be good to make the code shorter. Should it be done in this PR or another?

Reply:
If this change is used, there is no need of this PR anymore, right? No m or n is required. Maybe, create a separate PR (with the code and test) which would close this PR

Reply:
OK, but I think the comments specifying their dimensions could be maintained.

Reply:
By `(mat1.transpose() * mat2).trace()`, you are assuming both `mat1` and `mat2` are both `mxn`, right? So `mat1` don't need to be transposed before passed into `matricesInnerProd`. 

Should the function be changed to `(mat1.transpose() * mat2).trace()`? If so, `R`'s calculation in `df` and `fdf` should also be changed.

Reply:
The assumption isn't that both are `mxn`. The assumption would be:
```
mat1: MxN
mat2: MxP

mat1.transpose: NxM
mat1.transpose * mat2 = NxM * MxP = NxP
trace(above) = 1x1 (sum of min(N, P))
```

Reply:
I mean in the original implementation, `mat1` is already transposed outside `matricesInnerProd`, so it seems its equivalent should be `(mat1 * mat2).trace()`.
Also, should trace of non-square matrix be allowed here?

Reply:
Yeah, in that case, no transpose call. And trace of non-sq mat is well defined. No issues in allowing that